### ISSUE TITLE
set-add! sometimes causes program to hang, only when using CS

### ISSUE THREAD

**Issue by mosheduminer** [racket/racket#3738]
**What version of Racket are you using?**
8.0 [cs]

**What program did you run?**
```racket
#lang racket

(require racket/set)

(struct effect (dependencies) #:transparent #:mutable)
(define (create-effect-struct deps) (effect deps))

(define (subscribe running subscriptions)
  (set-add! subscriptions running)
  ; uh oh, this line causes the program to hang
  (set-add! (effect-dependencies running) subscriptions))

(define (test) (subscribe (create-effect-struct (mutable-set)) (mutable-set)) (writeln "success"))

(test)
```

**What should have happened?**
The program should execute, print "success", and exit. Instead, execution hangs on the second call to `set-add!`.
I could not figure out why the program stopped, and I decided to test using 8.0 BC. The program then ran and exited successfully. It seems this bug occurred because I was adding a struct containing a set to another set.

**Please include any other relevant details**
I am running on windows 10, using DrRacket.
I am a newcomer to racket, and was just playing around when I hit this bug. If this is in fact not a bug, or this is the wrong channel to report it, please direct me 🙂.

**Comment by sorawee** [comment]
The bug is caused by cyclic structure. Here's a minimal program that hangs:

```
(define s1 (mutable-set))
(define s2 (mutable-set))
(set-add! s1 s2)
(set-add! s2 s1)
```

**Comment by mflatt** [comment]
It's not exactly the cyclic structure, but the fact that there's a lock taken on `s` (at the level of a hash table) while it's being modified. During that modification, there's a need to inspect `s` to compute a hash code — but taking the lock for inspecting `s` gets stuck, because `s` is currently being modified.

I'm not immediately sure whether to call this a CS bug or a BC bug or both, given that BC isn't taking the lock on a hash table while it's computing a hash code. BC turns out to be unsafe, as illustrated by the program below that tends to crash on BC (and gets stuck on CS). Taking a lock would make it safe, but there are other ways to avoid a crash.

```
#lang racket/base

(define amok? #f)

(define ht (make-hash))

(struct a (x)
  #:property prop:equal+hash (list (lambda (a1 a2 eql?)
                                     (eql? (a-x a1) (a-x a2)))
                                   (lambda (a1 hc)
                                     (when amok?
                                       (hash-clear! ht))
                                     (a-x a1))
                                   (lambda (a2 hc)
                                     (a-x a2))))

(for ([i (in-range 1000)])
  (hash-set! ht (a i) #t))

(set! amok? #t)

(hash-set! ht ht #t)
```

**Comment by mflatt** [comment]
So far, I'm classifying this as a place to improve the documentation but not change the behavior of CS.

I've adjusted the "caveats concerning concurrent modification" for mutable hash tables to note the potential problem with adding a mutable hash table as part of a key in itself — already a problem, since adding to a hash table mutates it, and keys should not be mutated, but the symptom of getting stuck is a little different than the problem with mutating keys in general and worth noting. I also adjusted the documentation for operations like `set-add!` to point to the caveats for hash sets.

For the problems extra inspection uncovered: For BC, I've changed `hash-clear!` to take a the hash table's lock, and I've fixed problems in traversals that avoid taking the lock. For CS, I fixed problems in `hash-map` and `hash-for-each`, where they could crash as a result of concurrent modification.
